### PR TITLE
Add `AdditionalTextValueProvider<TValue>`

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Fixes https://github.com/dotnet/roslyn/issues/67611

This API acts the same as the existing SyntaxTreeValueProvider and SourceTextValueProvider.  These helpers allow diagnostic analzyers to cache data associated with a particular tree, 'source text', and now 'additional text', ensuring that they don't have to recompute that same data across analysis calls if the associated item is unchanged.

This helps analyzers avoid having to figure out how to do this themselves (including having to go down potentially problematic paths, like managing their own conditional weak tables).

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
@dotnet/roslyn-compiler ptal.  This was approved in last API review meeting.

**Comment by CyrusNajmabadi** [comment]
Ping @dotnet/roslyn-compiler .  Simple PR :)

File: src/Compilers/Core/Portable/PublicAPI.Unshipped.txt [hunk 1]
```diff
@@ -1,6 +1,11 @@
...
 *REMOVED*static Microsoft.CodeAnalysis.SyntaxList<TNode>.implicit operator Microsoft.CodeAnalysis.SyntaxList<TNode!>(Microsoft.CodeAnalysis.SyntaxList<Microsoft.CodeAnalysis.SyntaxNode!> nodes) -> Microsoft.CodeAnalysis.SyntaxList<TNode!>
 Microsoft.CodeAnalysis.Compilation.SupportsRuntimeCapability(Microsoft.CodeAnalysis.RuntimeCapability capability) -> bool
+Microsoft.CodeAnalysis.Diagnostics.AdditionalTextValueProvider<TValue>
+Microsoft.CodeAnalysis.Diagnostics.AdditionalTextValueProvider<TValue>.AdditionalTextValueProvider(System.Func<Microsoft.CodeAnalysis.AdditionalText!, TValue>! computeValue, System.Collections.Generic.IEqualityComparer<Microsoft.CodeAnalysis.AdditionalText!>? additionalTextComparer = null) -> void
+Microsoft.CodeAnalysis.Diagnostics.AnalysisContext.TryGetValue<TValue>(Microsoft.CodeAnalysis.AdditionalText! text, Microsoft.CodeAnalysis.Diagnostics.AdditionalTextValueProvider<TValue>! valueProvider, out TValue value) -> bool
```
Comment (cston) [inline src/Compilers/Core/Portable/PublicAPI.Unshipped.txt:6]: >AnalysisContext.TryGetValue<TValue>(Microsoft.CodeAnalysis.AdditionalText! text, Microsoft.CodeAnalysis.Diagnostics.AdditionalTextValueProvider<TValue>! valueProvider, out TValue value) [](http://example.com/codeflow?start=35&length=185)

It looks like this method was not included in the API review. Was that just an oversight?

**Review by cston — APPROVED** [review]

File: src/Compilers/Core/Portable/PublicAPI.Unshipped.txt (hunk 1 above)
Comment (CyrusNajmabadi) [reply src/Compilers/Core/Portable/PublicAPI.Unshipped.txt:6]: Yes.  The clear intent (and approval from api review) was simply that thsi mirror the existing SyntaxTree/SourceText providers, which this falls under.

**Pull request merged**